draft/XXXX-configuration_handling/README.md

```diff
@@ -0,0 +1,202 @@
+certain kind of "additive" effect as well. The idea here is to vary a set of
+parameters as a group. For example:
+
+```yaml
+a: 0
+b:
+  lockstep: foo
+  value: [1, 2, 3]
+c:
+  lockstep: foo
+  value: [4, 5, 6]
+```
+
+yielding the following `(a, b, c)` ordered triples:
+
+```
+(0, 1, 4)
+(0, 2, 5)
+(0, 3, 6)
+```
+
+Note that this requires specifying which "lockstep group" each parameter is a
+part of (identified by a string, in this case `foo`). Note also that `a = 0`
+once more, as it is specified as a plain value.
+
+#### Ranges
+
+We may want to support different types of range operators. The `value` field in
+the foregoing examples takes a YAML list as its argument, explicitly listing the
+values to take on. However, something like the following:
+
+```yaml
+a: 0
+b:
+  value:
+    sequence:
+      start: 1
+      stop: 3
+      step: 1
+c:
+  value:
+    sequence:
+      start: 4
+      stop: 6
+```
+
+would be the equivalent of the combinatorial example above (with the `step`
+parameter defaulting to `1`, as in the case of `c`).
+
+Other ranges can be implemented as needed.
+
+#### SMQTK algorithms
+
+Tinker Engine includes first-class support for SMQTK. A special
+meta-configuration option will enable SMQTK algorithms to be used as swappable
+values during configuration processing. For example:
+
+```yaml
+foo:
+  value:
+    - smqtk: MyAlgorithm.FooBar
+      config:
+        foo: 1
+        bar: 2
+    - smqtk: MyAlgorithm.FooBaz
+      config:
+        foo: 1
+        bar: 2
+```
+
+This will cause the `foo` value to take on a SMQTK algorithm instance as
+directed by the arguments. The runtime will first check to see that
+`MyAlgorithm.FooBar` and `MyAlgorithm.FooBaz` are existing, usable SMQTK
+algorithms as part of the validation phase.
+
+#### Schema and validation
+
+The meta-configuration schema will be formalized in such a way that the format
+can be presented in documentation easily, and so that configurations can be
+validated at runtime. A tool such as JSON Schema may be the right tool for this.
+
+#### Simple configuration
+
+As a special case, the meta-configuration can forgo any of these parameter
+varying options to produce a single configuration with all values statically
+set.
+
+### Programmatic Configuration
+
+As a fallback, the user can forgo using a declarative configuration at all and
+instead define a Python generator function that yields one or more simple
+configuration values itself. This could be useful if the declarative mechanisms
+are not expressive enough to achieve some result.
+
+As a matter of software architecture, the declarative meta-configuration logic
+will be implemented as a function of this type, that includes logic for parsing
+and interpreting the meta-configuration file, yielding concrete implementations
+as appropriate.
+
+## Reference Implementation
+
+Because this RFC is light on specific, concrete implementation direction, a
```

Comment:
One other seperate thing is GPU support.  We have had issues with being able to handle the GPU asked for by setting the CUDA_VISIBLE_DEVICES (or subsampling that if provided).  Something like this:  https://gitlab.kitware.com/darpa_learn/learn/-/blob/keep_last_stage_network/learn/protocol/localinterface.py#L68

Reply:
Is there something in the design of Tinker Engine that can help with this? It seems like an issue for a protocol writer. If there's something more concrete to track, could you open an issue in the appropriate repository? (This RFC doesn't seem to have anything to do with GPU support.)

Reply:
This would be useful to do this since it is by setting or updating the GPU environmental variable based on a GPU parameter.  Since this is just handling the GPU parameters, I see this as falling under configuration rather than GPU support in general.   

I'm proposing Tinker would be able to do the same as the linked code.  It will set CUDA_VISIBLE_DIVICES or augment it if it already define so that the protocol can be sent the correct GPUs.  Tinker would be a better place rather than each protocol having their own implementation of it (or worse, each algorithm).

Reply:
Are you saying you would want the Tinker Engine config file to have an `env` section where you could define runtime environment variables? We can do that. There are other tools that do this as well (for example, [dotenv](https://pypi.org/project/python-dotenv/)) but I can see how setting up those variables in the config file would be helpful, particularly for values that have to be set a certain way for a certain protocol setup.

Reply:
Or are you asking for a specific mechanism to help with GPU support? I need some details in order to get this into the RFC document.

Reply:
Let move this to another RFC